src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -843,15 +861,24 @@ private static unsafe void ClearImpl(Array array, int index, int length)
             if (array == null)
                 throw new ArgumentNullException("array");
 
+#if !REAL_MULTIDIM_ARRAYS
+            // NOTE: ONCE WE GET RID OF THE IFDEFS, WE SHOULD RENAME THIS METHOD.
+            // Get the backing array if this is an MDArray instance
             array = array.FlattenedArray;
+#endif
 
             if (index < 0 || index > array.Length || length < 0 || length > array.Length)
                 throw new IndexOutOfRangeException();
             if (length > (array.Length - index))
                 throw new IndexOutOfRangeException();
 
+#if REAL_MULTIDIM_ARRAYS
+            // The above checks should have covered all the reasons why Clear would fail.
+            Debug.Assert(false);
```

Comment:
This method is a weird performance optimization (it takes out throws from the Clear method).

* When REAL_MULTIDIM_ARRAYS is not defined, it handles the multidim array case, and all the failures.
* When REAL_MULTIDIM_ARRAYS is defined, it only handles the failure cases, because the runtime call in Clear should have handled multidim. It will only be called when Clear failed, and Clear can only fail if one of the inputs is wrong. That's why I have a TODO in the diff above to rename this method (ClearImpl right now) to something else.

Medium to long term, we're getting rid of the fake mdarrays and this method will probably just assert that one of the args is out of range and throw an IndexOutOfRangeException.